scripts/Diffraction/single_crystal/base_sx.py

```diff
@@ -0,0 +1,574 @@
+from typing import Sequence, Optional
+import numpy as np
+from enum import Enum
+import mantid.simpleapi as mantid
+from mantid.api import FunctionFactory, AnalysisDataService as ADS
+from FindGoniometerFromUB import getSignMaxAbsValInCol
+from mantid.geometry import CrystalStructure, SpaceGroupFactory, ReflectionGenerator, ReflectionConditionFilter
+from os import path
+
+from abc import ABC, abstractmethod
+
+
+class PEAK_TYPE(Enum):
+    FOUND = "found"
+    PREDICT = "predicted"
+    PREDICT_SAT = "predicted_sat"
+
+
+class INTEGRATION_TYPE(Enum):
+    MD = "int_MD"
+    MD_OPTIMAL_RADIUS = "int_MD_opt"
+    SKEW = "int_skew"
+
+
+class BaseSX(ABC):
+    def __init__(self, vanadium_runno: str):
+        self.runs = dict()
+        self.van_runno = vanadium_runno
+        self.van_ws = None
+        self.gonio_axes = None
+        self.sample_dict = None
+        self.n_mcevents = 1200
+
+    # --- decorator to apply to all runs is run=None ---
+    def default_apply_to_all_runs(func):
+        def apply(self, *args, **kwargs):
+            if "run" in kwargs and kwargs["run"] is not None:
+                func(self, *args, **kwargs)
+            else:
+                for run in self.runs.keys():
+                    func(self, *args, **kwargs, run=run)
+
+        return apply
+
+    # --- getters ---
+
+    def get_ws(self, run):
```

Comment:
Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

Reply:
> Does this return a ws name? Might be better to call it `get_ws_name` if yes. Similar comment for `get_md`

I would like to encourage users to use strings in their scripts (less chance of bugs through an object becoming invalidated, and easier/more transparent when assigning the output of an algorithm to have the same name) so I force the class setters to set the internal dict values with workspace names. That said the user can access the internal dict (a downside of python...) and set the workspace to a python object if they wanted - and it's designed so everything will (should!) work anyway! i.e. I assume the return from a getter or function input could be a string or python object  - hence the `BaseSX.retrieve` method is called a lot. This makes it a bit more cumbersome for the developer but I think should be more robust...

So I would say it is a getter of the dict field `ws` - but I wouldn't want to bake in an assumption as to the data type - but maybe that's just confusing? Or I should have both?

Reply:
Yep I've thought about this a bit more and definitely need to be clear what data type the getters return - I will force it to be a string and call put `name` in the function name. Not sure I will add getters for ws objects as well, will rely on BaseSX.retrieve if users want that for now (hopefully they won't need it now there's the helper functions)